Source/Shaders/ModelExperimental/MaterialStageFS.glsl

```diff
@@ -0,0 +1,219 @@
+{
+    czm_modelMaterial material;
+    material.diffuse = vec3(1.0);
+    material.specular = vec3(0.04); // dielectric (non-metal)
+    material.roughness = 0.001;
+    material.normal = vec3(0.0, 0.0, 1.0);
+    material.occlusion = 1.0;
+    material.emissive = vec3(0.0);
+    material.alpha = 1.0;
+    return material;
+}
+
+vec3 SRGBtoLINEAR3(vec3 srgbIn) 
+{
+    return pow(srgbIn, vec3(2.2));
+}
+
+vec4 SRGBtoLINEAR4(vec4 srgbIn) 
+{
+    vec3 linearOut = pow(srgbIn.rgb, vec3(2.2));
+    return vec4(linearOut, srgbIn.a);
+}
+
+vec2 computeTextureTransform(vec2 texCoord, mat3 textureTransform)
+{
+    return vec2(textureTransform * vec3(texCoord, 1.0));
+}
+
+#ifdef HAS_NORMALS
+vec3 computeNormal()
+{
+    vec3 ng = normalize(v_normal);
+
+    vec3 normal = ng;
+    #ifdef HAS_NORMAL_TEXTURE
+    vec2 normalTexCoords = TEXCOORD_NORMAL;
+        #ifdef HAS_NORMAL_TEXTURE_TRANSFORM
+        normalTexCoords = computeTextureTransform(normalTexCoords, u_normalTextureTransform);
+        #endif
+
+        #ifdef HAS_TANGENTS
+        // read tangents from varying
+        vec3 t = normalize(v_tangent.xyz);
+        vec3 b = normalize(cross(ng, t) * v_tangent.w);
+        mat3 tbn = mat3(t, b, ng);
+        vec3 n = texture2D(u_normalTexture, normalTexCoords).rgb;
+        normal = normalize(tbn * (2.0 * n - 1.0));
+        #elif defined(GL_OES_standard_derivatives)
+        // Compute tangents
+        vec3 pos_dx = dFdx(v_positionEC);
+        vec3 pos_dy = dFdy(v_positionEC);
+        vec3 tex_dx = dFdx(vec3(normalTexCoords,0.0));
+        vec3 tex_dy = dFdy(vec3(normalTexCoords,0.0));
+        vec3 t = (tex_dy.t * pos_dx - tex_dx.t * pos_dy) / (tex_dx.s * tex_dy.t - tex_dy.s * tex_dx.t);
+        t = normalize(t - ng * dot(ng, t));
+        vec3 b = normalize(cross(ng, t));
+        mat3 tbn = mat3(t, b, ng);
+        vec3 n = texture2D(u_normalTexture, normalTexCoords).rgb;
+        normal = normalize(tbn * (2.0 * n - 1.0));
+        #endif
+    #endif
+
+    return normal;
+}
+#endif
+
+czm_modelMaterial materialStage(czm_modelMaterial inputMaterial)
+{
+    czm_modelMaterial material = inputMaterial;
+
+    #ifdef HAS_NORMALS
+    material.normal = computeNormal();
+    #endif
+
+    vec4 baseColorWithAlpha = vec4(1.0);
+    // Regardless of whether we use PBR, set a base color
+    #ifdef HAS_BASE_COLOR_TEXTURE
+    vec2 baseColorTexCoords = TEXCOORD_BASE_COLOR;
+
+        #ifdef HAS_BASE_COLOR_TEXTURE_TRANSFORM
+        baseColorTexCoords = computeTextureTransform(baseColorTexCoords, u_baseColorTextureTransform);
+        #endif
+
+    baseColorWithAlpha = SRGBtoLINEAR4(texture2D(u_baseColorTexture, baseColorTexCoords));
+
+        #ifdef HAS_BASE_COLOR_FACTOR
+        baseColorWithAlpha *= u_baseColorFactor;
+        #endif
+    #elif defined(HAS_BASE_COLOR_FACTOR)
+    baseColorWithAlpha = u_baseColorFactor;
+    #endif
+
+    #ifdef HAS_COLOR_0
+    baseColorWithAlpha *= v_color_0;
+    #endif
```

Comment:
I still think this should multiply each set of vertex colors, but I'm ok for now, I'm mostly happy that texture coordinates are handled generically.